cpp/src/transpose/transpose.cu

```diff
@@ -0,0 +1,176 @@
+
+  for (size_type i = x; i < input.num_columns(); i += stride_x) {
+    for (size_type j = y; j < input.num_rows(); j += stride_y) {
+      output.column(j).element<T>(i) = input.column(i).element<T>(j);
+    }
+  }
+}
+
+/**
+ * @brief Transposes the null mask from ncols x nrows input to nrows x ncols output
+ * 
+ * @tparam T  Datatype of values pointed to by the views
+ * @param input[in]  Device view of input columns' data
+ * @param output[out]  Mutable device view of pre-allocated output columns' data
+ */
+__global__
+void gpu_transpose_valids(table_device_view const input, mutable_table_device_view output)
+{
+  size_type x = blockIdx.x * blockDim.x + threadIdx.x;
+  size_type y = blockIdx.y * blockDim.y + threadIdx.y;
+
+  size_type stride_x = blockDim.x * gridDim.x;
+  size_type stride_y = blockDim.y * gridDim.y;
+
+  size_type i = x;
+  auto active_threads = __ballot_sync(0xffffffff, i < input.num_columns());
+  while (i < input.num_columns()) {
+    for (size_type j = y; j < input.num_rows(); j += stride_y) {
+      auto const result = __ballot_sync(active_threads, input.column(i).is_valid(j));
+
+      // Only one thread writes output
+      if (0 == threadIdx.x % WARP_SIZE) {
+        output.column(j).set_mask_word(word_index(i), result);
+      }
+    }
+
+    // Update active threads before branching
+    i += stride_x;
+    active_threads = __ballot_sync(active_threads, i < input.num_columns());
+  }
+}
+
+struct launch_kernel {
+  template <typename T>
+  void operator()(
+    table_view const& input,
+    mutable_table_view& output,
+    bool has_null,
+    cudaStream_t stream)
+  {
+    // Copy input columns `data` and `valid` pointers to device
+    auto device_input = table_device_view::create(input, stream);
+    auto device_output = mutable_table_device_view::create(output, stream);
+
+    dim3 dimBlock(WARP_SIZE, WARP_SIZE);
+    dim3 dimGrid(std::min(util::div_rounding_up_safe(input.num_columns(), WARP_SIZE), MAX_GRID_SIZE),
+                 std::min(util::div_rounding_up_safe(input.num_rows(), WARP_SIZE), MAX_GRID_SIZE));
+
+    gpu_transpose<T><<<dimGrid, dimBlock, 0, stream>>>(*device_input, *device_output);
+
+    if (has_null) {
+      gpu_transpose_valids<<<dimGrid, dimBlock, 0, stream>>>(*device_input, *device_output);
+
+      // Force null counts to be recomputed next time they are queried
+      for (auto& column : output) {
+        column.set_null_count(UNKNOWN_NULL_COUNT);
+      }
+    }
+  }
+};
+
+}  // namespace
+
+namespace detail {
+
+std::unique_ptr<experimental::table> transpose(table_view const& input,
+  rmm::mr::device_memory_resource* mr, cudaStream_t stream)
+{
+  auto const input_ncols = input.num_columns();
+  auto const input_nrows = input.num_rows();
+
+  // If there are no rows in the input, return successfully
+  if (input_ncols == 0 || input_nrows == 0) {
+    return std::make_unique<experimental::table>(std::vector<std::unique_ptr<column>>{});
```

Comment:
You can just undelete the default ctor for `cudf::table`. I'm not sure why I deleted it in the first place.

Then you can just do:
```suggestion
    return std::make_unique<experimental::table>();
```

Reply:
Done